### ISSUE TITLE
Is it necessary to check the underlying code of a web page for WCAG testing?

### ISSUE THREAD

**Issue by Zoey0li** [w3c/wcag#4297]
When we are testing whether a certain web page complies with the WCAG standards, is it necessary to check the code of the web page? Can we just click on the components on the page? For example, when testing SC 2.1.1, can we check whether all the operable components can be operated normally by manually using the keyboard? That is to say, if all the functions of the web page can be completed by operating the web page components with the keyboard, then we can determine that this web page complies with WCAG SC 2.1.1. However, when testing SC 4.1.2, it seems that we really need to check the code of the web page to determine whether the web page complies with WCAG SC 4.1.2. So, is it impossible to complete all the WCAG tests without checking the code of the web page, right?

**Comment by patrickhlauke** [comment]
> is it necessary to check the code of the web page?

not necessarily, no. test that the outcomes of the various SCs are satisfied. where looking at the code can be helpful, though, is to verify that particularly for ARIA-related stuff things are done correctly, as screen readers often end up behaving very differently when encountering broken/incorrect ARIA ... and if you only test in one SR, you may not actually get a good idea if something's done right or not, as what you experience may be down to a quirk in your particular SR, or its own error-correction.

**Comment by TestPartners** [comment]
It depends on how accurate you want your results to be. Your results are likely to be very inaccurate if you don’t look at the code. All the testing tools, from single-purpose bookmarklets to insanely expensive products, have bugs. Screen readers are not a reliable way to test the success criteria, no matter how many of them you use.

Tools and assistive technologies can be helpful in finding issues that warrant investigation, but it would be negligent to just report what they say. To do a WCAG audit accurately, you MUST look at the code to verify whether what the tools and ATs are telling you is correct (it frequently isn’t). And you need to look at the code to find all the non-conformances they don’t tell you about. And you need to understand all the relevant coding paradigms such as responsive design, asynchronous loading, single page applications etc.

You can’t even do something as apparently simple as a colour contrast measurement because WCAG requires that you use the author-specified colours in the code, not the rendered colours on screen.

Even if you did do the testing without looking at the code, the results would not be particularly useful because you would not be able to say why the success criteria failed or recommend how they can be fixed. There may be some very junior roles where that’s acceptable, but I have never seen one and wouldn’t employ anyone to do it.

Steve Green
Managing Director
Test Partners Ltd

**Comment by bruce-usab** [comment]
@Zoey0li are you satisfied with the feedback you have received on your question?

**Comment by Zoey0li** [comment]
yes. thank all of you very much.
